File: files/mesoscale/examples/diffusion.py
"""Fourier diffusion lesson, related to Fourier_Julia.ipynb.
Run: python diffusion.py [output_directory]. Requires NumPy.
Periodic domain L=2*pi; D=1. Exact Fourier propagation, not time stepping.
"""
import sys
from pathlib import Path
import numpy as np

def run(output='diffusion_output'):
    out=Path(output);out.mkdir(parents=True,exist_ok=True)
    n=128;x=2*np.pi*np.arange(n)/n;k=2*np.pi*np.fft.fftfreq(n,d=2*np.pi/n)
    c0=1+.4*np.cos(x)+.2*np.cos(4*x);rows=[]
    for t in [0,.02,.1,.3,1.]:
        c=np.fft.ifft(np.fft.fft(c0)*np.exp(-k*k*t)).real
        exact=1+.4*np.exp(-t)*np.cos(x)+.2*np.exp(-16*t)*np.cos(4*x)
        rows.append([t,c.mean(),np.max(np.abs(c-exact))])
        np.savetxt(out/f'profile_{t:.2f}.csv',np.column_stack([x,c]),delimiter=',')
    np.savetxt(out/'diagnostics.csv',rows,delimiter=',',header='time,mean,max_exact_error',comments='')
    assert max(r[2] for r in rows)<1e-12
    print('Fourier diffusion: max exact error',max(r[2] for r in rows))
if __name__=='__main__':run(*sys.argv[1:2])

File: files/mesoscale/examples/diffusion.jl
# Fourier diffusion lesson related to Fourier_Julia.ipynb.
# julia diffusion.jl [output_directory]; requires FFTW.
using FFTW, Statistics, DelimitedFiles, Printf
function run(output="diffusion_output")
    mkpath(output);n=128;x=2π .* (0:n-1) ./ n;k=2π .* fftfreq(n,n/(2π))
    c0=1  .+ .4 .* cos.(x) .+ .2 .* cos.(4  .* x);rows=Vector{Float64}[]
    for t in [0.,.02,.1,.3,1.]
        c=real.(ifft(fft(c0) .* exp.(-k .^ 2  .* t)))
        exact=1  .+ .4*exp(-t) .* cos.(x) .+ .2*exp(-16*t) .* cos.(4  .* x)
        push!(rows,[t,mean(c),maximum(abs.(c .- exact))])
        writedlm(joinpath(output,@sprintf("profile_%.2f.csv",t)),hcat(x,c),',')
    end
    open(joinpath(output,"diagnostics.csv"),"w") do io
        println(io,"time,mean,max_exact_error");writedlm(io,permutedims(hcat(rows...)),',')
    end
    @assert maximum(r[3] for r in rows)<1e-12
    println("Fourier diffusion: max exact error ",maximum(r[3] for r in rows))
end
if abspath(PROGRAM_FILE) == @__FILE__
    run(ARGS...)
end
